src/wallet/test/coinselector_tests.cpp

```diff
@@ -0,0 +1,554 @@
+        wtx->nDebitCached = 1;
+    }
+    COutput output(wtx.get(), nInput, nAge, true /* spendable */, true /* solvable */, true /* safe */);
+    vCoins.push_back(output);
+    wtxn.emplace_back(std::move(wtx));
+}
+
+static void empty_wallet(void)
+{
+    vCoins.clear();
+    wtxn.clear();
+    balance = 0;
+}
+
+static bool equal_sets(CoinSet a, CoinSet b)
+{
+    std::pair<CoinSet::iterator, CoinSet::iterator> ret = mismatch(a.begin(), a.end(), b.begin());
+    return ret.first == a.end() && ret.second == b.end();
+}
+
+static long make_hard_case(int utxos, std::vector<CInputCoin>& utxo_pool)
+{
+    utxo_pool.clear();
+    long target = 0;
+    for (int i = 0; i < utxos; ++i) {
+        target += (long)1 << (utxos+i);
+        add_coin((long)1 << (utxos+i), 2*i, utxo_pool);
+        add_coin(((long)1 << (utxos+i)) + ((long)1 << (utxos-1-i)), 2*i + 1, utxo_pool);
+    }
+    return target;
+}
+
+// Branch and bound coin selection tests
+BOOST_AUTO_TEST_CASE(bnb_search_test)
+{
+
+    LOCK(testWallet.cs_wallet);
+
+    // Setup 
+    std::vector<CInputCoin> utxo_pool;
+    CoinSet selection;
+    CoinSet actual_selection;
+    CAmount value_ret = 0;
+    
+    /////////////////////////
+    // Known Outcome tests //
+    /////////////////////////
+    BOOST_TEST_MESSAGE("Testing known outcomes");
+
+    // Empty utxo pool
+    BOOST_CHECK(!SelectCoinsBnB(utxo_pool, 1 * CENT, 0.5 * CENT, selection, value_ret, nullptr));
+    selection.clear();
+    
+    // Add 1, 2, and 3, utxos
+    add_coin(1 * CENT, 1, utxo_pool);    
+    add_coin(2 * CENT, 2, utxo_pool);
+    add_coin(3 * CENT, 3, utxo_pool);
+    add_coin(4 * CENT, 4, utxo_pool);
+    
+    // Select 1 Cent
+    add_coin(1 * CENT, 1, actual_selection);
+    BOOST_CHECK(SelectCoinsBnB(utxo_pool, 1 * CENT, 0.5 * CENT, selection, value_ret, nullptr));
+    BOOST_CHECK(equal_sets(selection, actual_selection));
+    actual_selection.clear();
+    selection.clear();
+    
+    // Select 2 Cent
+    add_coin(2 * CENT, 2, actual_selection);
+    BOOST_CHECK(SelectCoinsBnB(utxo_pool, 2 * CENT, 0.5 * CENT, selection, value_ret, nullptr));
+    BOOST_CHECK(equal_sets(selection, actual_selection));
+    actual_selection.clear();
+    selection.clear();
+    
+    // Select 5 Cent
+    add_coin(4 * CENT, 4, actual_selection);
+    add_coin(1 * CENT, 1, actual_selection);
+    BOOST_CHECK(SelectCoinsBnB(utxo_pool, 5 * CENT, 0.5 * CENT, selection, value_ret, nullptr));
+    BOOST_CHECK(equal_sets(selection, actual_selection));
+    actual_selection.clear();
+    selection.clear();
+    
+    // Select 11 Cent, not possible
+    BOOST_CHECK(!SelectCoinsBnB(utxo_pool, 11 * CENT, 0.5 * CENT, selection, value_ret, nullptr));
+    actual_selection.clear();
+    selection.clear();
+    
+    // Select 10 Cent
+    add_coin(5 * CENT, 5, utxo_pool);
+    add_coin(5 * CENT, 5, actual_selection);
+    add_coin(4 * CENT, 4, actual_selection);
+    add_coin(1 * CENT, 1, actual_selection);
```

Comment:
It is forced to be include first in these tests so the solution is deterministic.